python/ray/experimental/workflow/workflow_storage.py

```diff
@@ -0,0 +1,229 @@
+    object_refs: Optional[List[str]] = None
+    # The workflows in the inputs of the workflow.
+    workflows: Optional[List[str]] = None
+
+    def is_recoverable(self) -> bool:
+        return (self.output_object_valid or self.output_step_id
+                or (self.args_valid and self.object_refs is not None
+                    and self.workflows is not None and self.func_body_valid))
+
+
+class WorkflowStorage:
+    """Access workflow in storage. This is a high-level function,
+    which does not care about the underlining storage implementation."""
+
+    def __init__(self,
+                 workflow_id: Optional[str] = None,
+                 store: Optional[storage.Storage] = None):
+        if workflow_id is None:
+            context = workflow_context.get_workflow_step_context()
+            workflow_id = context.workflow_id
+        if store is None:
+            store = storage.get_global_storage()
+        self._storage = store
+        self._workflow_id = workflow_id
+
+    def load_step_output(self, step_id: StepID) -> Any:
+        """Load the output of the workflow step from checkpoint.
+
+        Args:
+            step_id: ID of the workflow step.
+
+        Returns:
+            Output of the workflow step.
+        """
+        return self._storage.load_step_output(self._workflow_id, step_id)
+
+    def load_step_func_body(self, step_id: StepID) -> Callable:
+        """Load the function body of the workflow step.
+
+        Args:
+            step_id: ID of the workflow step.
+
+        Returns:
+            A callable function.
+        """
+        return self._storage.load_step_func_body(self._workflow_id, step_id)
+
+    def load_step_args(
+            self, step_id: StepID, workflows: List[Any],
+            object_refs: List[ray.ObjectRef]) -> Tuple[List, Dict[str, Any]]:
+        """Load the input arguments of the workflow step. This must be
+        done under a serialization context, otherwise the arguments would
+        not be reconstructed successfully.
+
+        Args:
+            step_id: ID of the workflow step.
+            workflows: The workflows in the original arguments,
+                replaced by the actual workflow outputs.
+            object_refs: The object refs in the original arguments.
+
+        Returns:
+            Args and kwargs.
+        """
+        with serialization_context.workflow_args_resolving_context(
+                workflows, object_refs):
+            return self._storage.load_step_args(self._workflow_id, step_id)
+
+    def load_object_ref(self, object_id: str) -> ray.ObjectRef:
+        """Load the input object ref.
+
+        Args:
+            object_id: The hex ObjectID.
+
+        Returns:
+            The object ref.
+        """
+        return self._storage.load_object_ref(self._workflow_id, object_id)
+
+    def _update_output_shortcut(self, forward_output_to: StepID,
+                                shortcut_output_step_id: StepID) -> None:
```

Comment:
For clarity, can we instead call these

`original_output_step_id` and `dynamic_output_step_id`? Then it becomes much clearer to explain that in dynamic DAGs, the output step id can change and dynamic_output_step id caches the latest step so we don't need to traverse the dynamically generated DAG to find the actual step producing the output.